aisdc/attacks/attack_report_formatter.py

```diff
@@ -170,22 +232,37 @@ def __init__(self, report: dict, metrics_list=None):
         self.report = report
         self.metrics_list = metrics_list
 
-    def process_dict(self) -> dict:
-        metrics_dict = {m: [] for m in self.metrics_list}
-        for _, iteration_value in self.report["attack_experiment_logger"][
-            "attack_instance_logger"
-        ].items():
-            for m in metrics_dict:
-                metrics_dict[m].append(iteration_value[m])
-        output = {}
-        for m in self.metrics_list:
-            output[m] = {
-                "min": min(metrics_dict[m]),
-                "max": max(metrics_dict[m]),
-                "mean": np.mean(metrics_dict[m]),
-                "median": np.median(metrics_dict[m]),
-            }
-        return output
+        self.immediate_rejection = []
+        self.support_rejection = []
+        self.support_release = []
+
+    def process_dict(self):
+        output_dict = {}
+
+        for k in self.report.keys():
+            if isinstance(self.report[k], dict):
+                if "attack_experiment_logger" in self.report[k]:
+                    metrics_dict = {m: [] for m in self.metrics_list}
+                    for _, iteration_value in self.report[k][
+                        "attack_experiment_logger"
+                    ]["attack_instance_logger"].items():
+                        for m in metrics_dict:
+                            metrics_dict[m].append(iteration_value[m])
+                    output = {}
+                    for m in self.metrics_list:
+                        output[m] = {
+                            "min": min(metrics_dict[m]),
+                            "max": max(metrics_dict[m]),
+                            "mean": np.mean(metrics_dict[m]),
+                            "median": np.median(metrics_dict[m]),
+                        }
+                    output_dict[k] = output
+        return (
+            output_dict,
+            self.immediate_rejection,
+            self.support_rejection,
+            self.support_release,
```

Comment:
In a class FinalRecommendationsModule, a method process_dict includes returning variables like immediate_rejections, support_rejection and support_release are updated and then returned.  Whereas in classes like  SummariseUnivariateMetricsModule, SummariseAUCPValsModule and LogLogROCModule, for process_dict method these variables are not updated and seem written just to maintain the returning variables pattern align with the process_diction method of the FinalRecommendationonly class. I wonder if could we think of some other way like when all these classes are inheriting a class AnalysisModule, should we not define these variables in the superclass and use them where required, and this will avoid repeating them across multiple classes?

Reply:
No that makes perfect sense - I spotted the same issue when I was writing the code, and I didn't like what I was writing, but I couldn't see a neat solution. This is exactly the solution I was looking for - thanks!